File: R/transport.R
# Encode data frames compactly for the private R-to-JavaScript widget payload.
# Public GenomeSpy specifications remain unchanged; the browser reconstructs
# ordinary row objects before handing the specification to GenomeSpy.
mutglyph_widget_to_json <- function(x, pretty = FALSE) {
  mutglyph_to_json(mutglyph_encode_transport(x), pretty = pretty)
}

mutglyph_encode_transport <- function(x) {
  if (is.data.frame(x)) {
    return(mutglyph_encode_data_frame(x))
  }
  if (is.list(x)) {
    return(lapply(x, mutglyph_encode_transport))
  }
  x
}

mutglyph_encode_data_frame <- function(data) {
  list(
    `$type` = "mutglyph-data-frame",
    rows = nrow(data),
    names = unname(as.list(names(data))),
    columns = unname(lapply(data, mutglyph_encode_column))
  )
}

mutglyph_encode_column <- function(column) {
  if (is.factor(column)) {
    column <- as.character(column)
  }

  if (!is.character(column) || length(column) == 0L) {
    # Keep atomic vectors intact. Converting every value to a one-element R
    # list makes jsonlite walk large columns item by item.
    return(mutglyph_transport_array(column))
  }

  present <- !is.na(column)
  present_count <- sum(present)
  if (present_count == 0L || length(column) < 32L) {
    return(mutglyph_transport_array(column))
  }

  dictionary <- unique(column[present])
  # A dictionary is useful only for clearly repetitive values. This cheap
  # cardinality test avoids serializing both candidate representations just to
  # compare their sizes, which is costly for large annotation tracks.
  if (length(dictionary) > 0.5 * present_count) {
    return(mutglyph_transport_array(column))
  }

  codes <- match(column, dictionary) - 1L
  codes[!present] <- NA_integer_
  encoded <- list(
    dictionary = mutglyph_transport_array(dictionary),
    codes = mutglyph_transport_array(codes)
  )
  encoded
}

mutglyph_transport_array <- function(x) {
  # auto_unbox would turn a one-element vector into a scalar, but the browser
  # decoder requires every transport column and dictionary to be an array.
  if (length(x) == 1L) as.list(x) else unname(x)
}

File: srcjs/transport.js
const DATA_FRAME_TYPE = "mutglyph-data-frame";

/**
 * Reconstructs ordinary row objects from MutGlyph's private columnar widget
 * payload. GenomeSpy and downloaded specifications never see this wire format.
 */
export function decodeMutGlyphTransport(value) {
  if (Array.isArray(value)) {
    return value.map(decodeMutGlyphTransport);
  }
  if (value === null || typeof value !== "object") {
    return value;
  }
  if (value.$type === DATA_FRAME_TYPE) {
    return decodeDataFrame(value);
  }

  return Object.fromEntries(
    Object.entries(value).map(([key, child]) => [
      key,
      decodeMutGlyphTransport(child),
    ]),
  );
}

function decodeDataFrame(value) {
  const { rows, names, columns } = value;
  if (
    !Number.isInteger(rows) ||
    rows < 0 ||
    !Array.isArray(names) ||
    !Array.isArray(columns) ||
    names.length !== columns.length
  ) {
    throw new Error("Invalid MutGlyph data-frame payload.");
  }

  const decodedColumns = columns.map((column) => decodeColumn(column, rows));
  return Array.from({ length: rows }, (_, rowIndex) =>
    Object.fromEntries(
      names.map((name, columnIndex) => [
        name,
        decodedColumns[columnIndex][rowIndex],
      ]),
    ),
  );
}

function decodeColumn(column, rows) {
  let values;
  if (Array.isArray(column)) {
    values = column.map(decodeMutGlyphTransport);
  } else if (
    column !== null &&
    typeof column === "object" &&
    Array.isArray(column.dictionary) &&
    Array.isArray(column.codes)
  ) {
    const dictionary = column.dictionary.map(decodeMutGlyphTransport);
    values = column.codes.map((code) =>
      code === null ? null : dictionary[code],
    );
  } else {
    throw new Error("Invalid MutGlyph data-frame column.");
  }

  if (values.length !== rows) {
    throw new Error("MutGlyph data-frame column length does not match its rows.");
  }
  return values;
}
